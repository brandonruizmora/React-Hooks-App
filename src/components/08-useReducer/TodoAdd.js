import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddNewTodo }) => {

    const [{ description }, handleInputChange, reset] = useForm({
        description: ''
    });

    const handleSubmit = function (e) {
        e.preventDefault();

        if (description.trim().length <= 1) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        handleAddNewTodo(newTodo);

        reset();
    }

    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Hacer..."
                    onChange={handleInputChange}
                    value={description}
                    autoComplete="off"
                />
                <div className="d-grid">
                    <button
                        type="submit"
                        className="btn btn-outline-primary mt-2"
                    >
                        Agregar
                    </button>
                </div>
            </form>
        </>
    )
}
