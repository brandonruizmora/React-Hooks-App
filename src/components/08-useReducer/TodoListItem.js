import React from 'react';
import PropTypes from 'prop-types';

export const TodoListItem = ({ todo, index, handleToggle, handleDelete }) => {
    return (
        <li
            className="list-group-item"
        >
            <div className="d-flex align-items-center justify-content-between">
                <span
                    onClick={() => handleToggle(todo.id)}
                    className={todo.done ? "pointer text-decoration-line-through" : "pointer"}
                >
                    {index + 1}. {todo.desc}
                </span>
                <button
                    className="btn btn-danger"
                    onClick={() => handleDelete(todo.id)}
                >
                    borrar
                </button>
            </div>
        </li>
    )
}

TodoListItem.propTypes = {
    todo: PropTypes.object.isRequired,
    index: PropTypes.number.isRequired,
    handleToggle: PropTypes.func.isRequired,
    handleDelete: PropTypes.func.isRequired
}
