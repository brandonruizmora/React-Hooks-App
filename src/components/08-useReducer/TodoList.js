import React from 'react';
import PropTypes from 'prop-types';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({ todos, handleToggle, handleDelete }) => {
    return (
        <ul className="list-group list-group-flush">
            {
                todos.map((todo, idx) => {
                    return (
                        <TodoListItem
                            key={todo.id}
                            todo={todo}
                            index={idx}
                            handleToggle={handleToggle}
                            handleDelete={handleDelete}
                        />
                    )
                })
            }
        </ul>
    )
}

TodoList.propTypes = {
    todos: PropTypes.array.isRequired,
    handleToggle: PropTypes.func.isRequired,
    handleDelete: PropTypes.func.isRequired
}
