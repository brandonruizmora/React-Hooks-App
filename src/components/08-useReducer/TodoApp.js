import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './TodoReducer';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import './TodoApp.css';

export const TodoApp = () => {

    const init = function () {
        return JSON.parse(localStorage.getItem('todos')) || [];
    }

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos])

    const handleDelete = function (todoId) {

        const action = {
            type: 'delete',
            payload: todoId
        }

        dispatch(action);

    }

    const handleToggle = function (todoId) {
        const action = {
            type: 'toggle',
            payload: todoId
        }
        dispatch(action);
    }

    const handleAddNewTodo = function (newTodo) {
        const action = {
            type: 'add',
            payload: newTodo
        };
        dispatch(action);
    }

    return (
        <div className="container">
            <h1>TodoApp ({todos.length})</h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    <TodoList
                        todos={todos}
                        handleToggle={handleToggle}
                        handleDelete={handleDelete}
                    />
                </div>
                <div className="col-5">
                    <TodoAdd handleAddNewTodo={handleAddNewTodo} />
                </div>

            </div>
        </div>
    )
}
