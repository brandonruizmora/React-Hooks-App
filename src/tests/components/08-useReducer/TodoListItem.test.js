import '../../../setupTest';
import { shallow } from 'enzyme';
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Pruebas en <TodoListItem />', () => {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow(
        <TodoListItem
            todo={demoTodos[0]}
            index={0}
            handleToggle={ handleToggle }
            handleDelete={ handleDelete }
        />
    );

    test('Debe de renderizar el componente correctamente', () => {

        expect(wrapper).toMatchSnapshot();

    });

    test('Debe llamar la funcion handleDelete', () => {
        wrapper.find('button').simulate('click');
        expect( handleDelete ).toHaveBeenCalledWith( demoTodos[0].id );
    });

    test('Debe llamar la funcion handleToggle', () => {
        wrapper.find('span').simulate('click');
        expect( handleToggle ).toHaveBeenCalledWith( demoTodos[0].id );
    });

    test('Debe de mostrar el texto correctamente', () => {
        expect( wrapper.find('span').text() ).toBe(`1. ${demoTodos[0].desc}`);
    });

    test('Debe de tener la clase complete si el TODO.done = true', () => {

        const todo = demoTodos[0];
        todo.done = true;

        const wrapper = shallow(
            <TodoListItem
                todo={todo}
                index={0}
                handleToggle={ handleToggle }
                handleDelete={ handleDelete }
            />
        );

        expect( wrapper.find('span').hasClass('text-decoration-line-through') ).toBe( true );

    });

});
