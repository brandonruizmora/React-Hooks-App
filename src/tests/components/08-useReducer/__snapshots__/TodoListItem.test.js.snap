// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoListItem /> Debe de renderizar el componente correctamente 1`] = `
<li
  className="list-group-item"
>
  <div
    className="d-flex align-items-center justify-content-between"
  >
    <span
      className="pointer"
      onClick={[Function]}
    >
      1
      . 
      Aprender React
    </span>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      borrar
    </button>
  </div>
</li>
`;
